src/analysis/souffle/taint.dl

```diff
@@ -71,6 +92,32 @@ mayHaveTag(tgt, owner, tag) :- hasTag(tgt, owner, tag).
 mayHaveTag(tgt, owner, tag) :-
     resolvedEdge(owner, src, tgt), mayHaveTag(src, owner, tag), !removeTag(tgt, owner, tag).
 
+// Integrity tag rules.
+isIntegrityTag(integTag) :- directHasIntegrityTag(_, _, integTag).
+isPrincipal(principal) :- directHasIntegrityTag(_, principal, _).
+
+// Any access point that has no incoming edges and is not directly given an
+// integrity tag lacks that integrity tag, as there is no way for it to
+// propagate to that AccessPath.
+lacksIntegrityTag(ap, owner, integTag) :-
+  isAccessPath(ap), isPrincipal(owner), isIntegrityTag(integTag),
+  !edge(_, ap), !directHasIntegrityTag(ap, owner, integTag).
+
+// Any point where an integrity tag is removed lacks an integrity tag.
+lacksIntegrityTag(ap, owner, integTag) :- isAccessPath(ap), removeIntegrityTag(ap, owner, integTag).
+
+// Propagate lack of integrity tag along edges to any destination that did not
+// have an integrity tag directly added.
+lacksIntegrityTag(dst, owner, integTag) :-
+  isAccessPath(src), isAccessPath(dst),
+  edge(src, dst), lacksIntegrityTag(src, owner, integTag),
+  !directHasIntegrityTag(dst, owner, integTag).
+
+// An access path must have an integrity tag if it does not lack the integrity tag.
+mustHaveIntegrityTag(ap, owner, integTag) :-
```

Comment:
This still has the problem that `directHasIntegrity` does not imply `mustHaveIntegrityTag`.

Reply:
Sorry, I missed your comment in the other thread because I only looked at the `Files Changed` tab. Looks like this does work.